test/support/protocol_utils.rb

```diff
@@ -0,0 +1,343 @@
+      send_request 'stackTrace',
+                    threadId: 1,
+                    startFrame: 0,
+                    levels: 20
+      res = find_crt_response
+      f_id = res.dig(:body, :stackFrames, frame_idx, :id)
+      send_request 'evaluate',
+                    expression: expression,
+                    frameId: f_id,
+                    context: context
+      res = find_crt_response
+      assert_dap_res :EvaluateResponse, res
+
+      failure_msg = FailureMessage.new{create_protocol_message "result:\n#{JSON.pretty_generate res}"}
+      if expected.is_a? String
+        expected_val = expected
+      else
+        expected_val = expected.inspect
+      end
+      result_val = res.dig(:body, :result)
+      assert_equal expected_val, result_val, failure_msg
+
+      expected_type = expected.class.inspect
+      result_type = res.dig(:body, :type)
+      assert_equal expected_type, result_type, failure_msg
+    end
+
+    def send_request command, **kw
+      send type: 'request',
+            command: command,
+            arguments: kw
+    end
+
+    def send **kw
+      kw[:seq] = @seq += 1
+      str = JSON.dump(kw)
+      @sock.write "Content-Length: #{str.bytesize}\r\n\r\n#{str}"
+      @backlog << "V>D #{str}"
+    end
+
+    TIMEOUT_SEC = (ENV['RUBY_DEBUG_TIMEOUT_SEC'] || 10).to_i
+
+    def assert_dap_res expected_def, result_res
+      properties = @protocol.dig(:definitions, expected_def, :allOf, 1)
+      failure_msg = FailureMessage.new{create_protocol_message "expected:\n#{JSON.pretty_generate properties}\n\nresult:\n#{JSON.pretty_generate result_res}"}
+      patterns = DAP_ResponsePattern.new.get properties
+      patterns.each{|keys, expected|
+        result = result_res.dig(*keys)
+        assert_kind_of expected, result, failure_msg
+      }
+    end
+
+    def fetch_dap_json
+      return if File.exist? DAP_JSON_PATH
+
+      json = Net::HTTP.get(URI.parse('https://microsoft.github.io/debug-adapter-protocol/debugAdapterProtocol.json'))
+      File.write DAP_JSON_PATH, json
+    end
+
+    def find_crt_response
+      Timeout.timeout(TIMEOUT_SEC) do
+        loop do
+          res = @queue.pop
+          str = JSON.dump(res)
+          @backlog << "V<D #{str}"
+          if res[:request_seq] == @seq
+            return res
+          end
+        end
+      end
+    rescue Timeout::Error
+      flunk create_protocol_message "TIMEOUT ERROR (#{TIMEOUT_SEC} sec) while waiting seq: #{@seq}"
+    end
+
+    # FIXME: Commonalize this method.
+    def create_protocol_message fail_msg
+      all_protocol_msg = <<~DEBUGGER_MSG.chomp
+        -------------------------
+        | All Protocol Messages |
+        -------------------------
+
+        #{@backlog.join("\n")}
+      DEBUGGER_MSG
+
+      last_msg = @backlog.reverse[0..3].map{|m|
+        h = m.sub(/(D|V)(<|>)(D|V)\s/, '')
+        JSON.pretty_generate(JSON.parse h)
+      }.reverse.join("\n")
```

Comment:
I feel that this part may need a comment for explanation.

Reply:
Thanks, this part is complex.
I'll refactor the logic.